### DISCUSSION TITLE
Own YOLOX model doesn't work

### DISCUSSION THREAD

**Discussion by OlegKochetkov-code** [asus4/onnxruntime-unity-examples#14] [Q&A]
Hello. I trained a YOLOX model to recognize hands, but it did not work in your plugin. In python inference runs fine. I changed the model itself, the list of classes, the number of classes. Maybe something else needs to be done? In other words, what needs to be done to make your own YOLOX model work? Your example works fine https://github.com/asus4/onnxruntime-unity-examples

**Comment by asus4** [comment]
Hi @OlegKochetkov-git can you just upload your reproducible code and the model on GitHub

**Reply by OlegKochetkov-code** [reply]
https://github.com/OlegKochetkov-git/onnxruntime-unity-examples - Here's an example of my hand model that doesn't work. But if you change the frame, as it was with the COCO model, change the labels and the number of classes, everything works.

https://github.com/OlegKochetkov-git/YOLOX - Here is the model I trained. The model will run on the processor in the future.
[Link](https://drive.google.com/file/d/1VRiqZUkdzQfkST--BMP0eQYDc0l-GRxN/view?usp=sharing) to mini dataset. These folders should be placed [here](https://github.com/OlegKochetkov-git/YOLOX/tree/main/datasets)
[Link](https://drive.google.com/file/d/1UsBx3At8A0753_oMhH-SqI7YiXVBgif6/view?usp=sharing) to weight, onnx, ort file which I received. You can check them by placing them in the root of the model project and running [these](https://github.com/OlegKochetkov-git/YOLOX/blob/main/command_for_start.txt) commands.
The model is not trained perfectly, but I just wanted to check the functionality of your plugin. Thanks in advance.

**Reply by asus4** [reply]
@OlegKochetkov-git Hmm, i didn't find any issue in C#. 


I'm wondering if the input range in your model might be (0 ~ 255) by simply casting uint8 to float32?. (Not tested)

https://github.com/OlegKochetkov-git/YOLOX/blob/4e4ee91181ed459aa2cde3568ee1dc6d2499d73a/yolox/data/data_augment.py#L142-L158

**Reply by OlegKochetkov-code** [reply]
@asus4 I changed everything in this function to float32 instead of uint8, the model's inference shows normal results. Those. don't change anything.
Another question. Was it meant that any trained YOLOX model would work on Unity without problems?

**Reply by asus4** [reply]
No, I confirmed with only the pre-trained model which I'm using. Even if the model works on ONNXRuntime, the pre/post-processing can be slightly different in each model.

**Reply by OlegKochetkov-code** [reply]
@asus4 A possible solution would be to rewrite the preprocessing from Python to C#? Take this [code](https://github.com/OlegKochetkov-git/YOLOX/blob/4e4ee91181ed459aa2cde3568ee1dc6d2499d73a/yolox/data/data_augment.py#L142-L158) rewrite it in Yolox [class](https://github.com/OlegKochetkov-git/onnxruntime-unity-examples/blob/main/Assets/Samples/Yolox/Yolox.cs) in ovveride method PreProcess? PostPorcessing should not differ from your implementation.

**Reply by asus4** [reply]
Yes, it's a possible solution. I don't confirm that the issue is in the pre-processing though.

**Reply by OlegKochetkov-code** [reply]
OK, thank you. If something works out, I’ll post it in case it’s useful to someone.

**Reply by OlegKochetkov-code** [reply]
@asus4 Hi again. How to make sure that the model runs on the windows CPU? And where did you get [this](https://github.com/asus4/onnxruntime-unity-examples/tree/main/Assets/Samples/Yolox) model -yolox_nano_320x320.with_runtime_opt.ort?

**Reply by PauloBrito-DTX** [reply]
@OlegKochetkov-code @asus4 Hi, I think I have the same issue. I trained a model in yolox for digit recognition, it works in python, but when I convert to ONNX->ORT doesn't work in unity. Also I tested yolox_nano and yolox_s (converted to ONNX->ORT) in a working scene in unity instead of yolox_nano_320x320, and it didn't work. 
The difference between yolox_nano_320x320 and yolox_nano is just the input size? (Input shape: 1,3,320,320 to 1,3,416,416)
Do you have any solution?
Thks

**Reply by asus4** [reply]
Hi @PauloBrito-DTX, Sorry for the late reply @OlegKochetkov-code 

The included ORT model was initially converted from here.
https://github.com/asus4/onnxruntime-unity-examples/blob/dc13d52103a3b52ceea4d1451b1ce88b7005c5bb/Assets/Samples/Yolox/Yolox.cs#L20

The model's input/output looks different from the original one. I guess there are some updates to the latest model. I will modify the Yolox example using the original onnx model.
https://github.com/Megvii-BaseDetection/YOLOX/tree/main/demo/ONNXRuntime

![Screenshot 2025-01-29 at 11 47 50 Large](https://github.com/user-attachments/assets/c6be0f71-dc12-4775-aa81-ebdb252da8ff)

**Reply by asus4** [reply]
Hi @OlegKochetkov-code @PauloBrito-DTX, 

I figured out the cause of this issue. The post-process was correct, but the input range in the pre-process was different. It worked as expected when I changed the input range from `0.0~1.0` to `0~255`.
 
changes in ImageProcessing options 

- Mean: `0`
- StdDev: `1/255 = 0.003921569`

![Screenshot 2025-01-29 at 15 56 55](https://github.com/user-attachments/assets/f284f87a-33bb-4c21-8947-a1a1e1646066)

I will update the model to the original one in this example to avoid such confusion.

**Reply by asus4** [reply]
Hi, I replaced the model with the original version in PR #22. I hope your custom model works with this configuration🤞

**Comment by PauloBrito-DTX** [comment]
So i tested this:
python demo/ONNXRuntime/onnx_inference.py -m yolox_nano.with_runtime_opt.ort -i assets/dog.jpg -o YOLOX_outputs/yolox_nano/vis_res -s 0.25 --input_shape 416,416
and got this result:
![dog](https://github.com/user-attachments/assets/7cacc9ba-578e-42a2-b880-2cce42ff5007)
The same test in yolox_nano_320x320:
python demo/ONNXRuntime/onnx_inference.py -m yolox_nano_320x320.with_runtime_opt.ort -i assets/dog.jpg -o YOLOX_outputs/yolox_nano/vis_res -s 0.25 --input_shape 320,320
Result:
![dog](https://github.com/user-attachments/assets/8f368e8d-9ca7-4853-96d0-464c8bf09421)
So I will look into preprocess in onnx_inference.py.

**Reply by asus4** [reply]
Thanks @PauloBrito-DTX , the postprocess may be difference from the latest original Yolox. We need to update the postprocess.
